File: shoring/blackjackEngine/src/main/java/com/panopset/blackjackEngine/Cycle.java
package com.panopset.blackjackEngine;

import static com.panopset.blackjackEngine.CommandDefinitions.CMD_DEAL;
import java.io.StringWriter;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

public class Cycle {

  Cycle(final BlackjackGameEngine game, final Shoe shoe, final Strategy strategy) {
    this.strategy = strategy;
    blackjackConfiguration = game.getConfig();
    msg = game.getConfig().getMessages();
    bge = game;
    this.shoe = shoe;
  }

  @Override
  public String toString() {
    StringWriter sw = new StringWriter();
    sw.append("Dealer: ");
    if (dealer != null) {
      sw.append(String.format("%s%n", dealer.cards));
    }
    int i = 0;
    for (Player p : getPlayers()) {
      sw.append(String.format("Player %d: %s%n", i++, p.toString()));
    }
    sw.append(" Bankroll:" + bge.getBankroll());
    return sw.toString();
  }

  private final BlackjackConfiguration blackjackConfiguration;
  private final BlackjackMessages msg;

  private final BlackjackGameEngine bge;

  private final Shoe shoe;

  private final Strategy strategy;

  private HandDealer dealer;

  private List<Player> players;

  public HandDealer getDealer() {
    if (dealer == null) {
      dealer = new HandDealer();
    }
    return dealer;
  }

  public List<Player> getPlayers() {
    if (players == null) {
      players = new CopyOnWriteArrayList<>();
      for (int i = 0; i < blackjackConfiguration.getSeats(); i++) {
        if (blackjackConfiguration.isCountVeryNegative()) {
          bge.triggerShuffleBeforeNextHand();
        }
        Player player = new Player(bge);
        int betAmount = new BetAmountStrategy(bge).adjust();
        player.getHands().get(0).getWager().setInitialBet(betAmount);
        players.add(player);
      }
    }
    return players;
  }

  public Player getActivePlayer() {
    for (Player p : getPlayers()) {
      if (!p.isFinal()) {
        return p;
      }
    }
    return null;
  }

  public String getRecommendedAction() {
    if (getDealer().isFinal() || !getDealer().hasCards()) {
      return CMD_DEAL;
    } else {
      return getStrategy().getRecommendation(getCurrentSituation(getActivePlayerHand()));
    }
  }

  public StrategyLine getStrategyLine() {
    if (getActivePlayer() == null) {
      return null;
    }
    return getStrategy().findStrategyLine(getCurrentSituation(getActivePlayer().getActiveHand()));
  }

  public Situation getActiveSituation() {
    return getCurrentSituation(getActivePlayerHand());
  }

  public Situation getCurrentSituation(HandPlayer hand) {
    if (isDealt()) {
      return new Situation(getDealer().getUpCard(), hand);
    }
    return new Situation(null, hand);
  }

  public Strategy getStrategy() {
    return strategy;
  }

  public void deal() {
    for (Player p : getPlayers()) {
      p.getActiveHand().dealCard(new BlackjackCard(shoe.deal()));
      bge.reportNewHand();
    }
    getDealer().dealCard(new BlackjackCard(shoe.deal(), false));
    for (Player p : getPlayers()) {
      p.getActiveHand().dealCard(new BlackjackCard(shoe.deal()));
    }
    getDealer().dealCard(new BlackjackCard(shoe.deal()));
    checkFor21();
    dealt = true;
    active = true;
  }

  private void checkFor21() {
    for (Player player : getPlayers()) {
      checkFor21(player);
    }
    if (getDealer().getValue() == 21) {
      finish();
    }
  }

  private void checkFor21(Player player) {
    if (getDealer().getUpCard().isAce() && bge.getConfig().isEvenMoneyOnBlackjackVace()) {
      for (HandPlayer playerHand : player.getHands()) {
        if (playerHand.isNatural21()) {
          playerHand.standWithEvenMoney();
        }
      }
    } else {
      for (HandPlayer playerHand : player.getHands()) {
        if (playerHand.getValue() == 21) {
          player.getActiveHand().stand();
        }
      }
    }
  }

  public HandPlayer getActivePlayerHand() {
    Player p = getActivePlayer();
    if (p == null) {
      return null;
    }
    return p.getActiveHand();
  }

  public synchronized void finish() {
    synchronized (bge) {
      completeCycle();
      active = false;
    }
  }

  private void completeCycle() {
    if (!getDealer().isFinal()) {
      prepareSettlement();
    }
  }

  private void prepareSettlement() {
    getDealer().setShowing(0);
    resolvePlayerBlackjacks();
    while (!getDealer().isFinal()) {
      if (getDealer().getValue() < 17) {
        getDealer().dealCard(bge.deal());
      } else if (getDealer().getValue() == 17) {
        if (blackjackConfiguration.isDealerHitSoft17() && getDealer().isSoft()) {
          getDealer().dealCard(bge.deal());
        } else {
          getDealer().stand();
        }
      } else {
        getDealer().stand();
      }
    }
    bge.clearPriorHandBustedFlag();
    collectAndPayChips();
  }

  private void resolvePlayerBlackjacks() {
    boolean allPlayersHaveBlackjack = true;
    for (Player handPlayer : getPlayers()) {
      if (!resolvePlayerHandsBlackjacks(handPlayer)) {
        allPlayersHaveBlackjack = false;
      }
    }
    if (allPlayersHaveBlackjack) {
      dealer.stand();
    }
  }

  private boolean resolvePlayerHandsBlackjacks(Player handPlayer) {
    boolean allHandsAreBlackjack = true;
    for (HandPlayer h : handPlayer.getHands()) {
      if (!h.isDone()) {
        if (getDealer().isNatural21() && !blackjackConfiguration.isEuropeanStyle()) {
          h.stand();
        } else {
          if (h.isNatural21()) {
            h.stand();
          } else {
            allHandsAreBlackjack = false;
          }
        }
      }
    }
    return allHandsAreBlackjack;
  }

  private void collectAndPayChips() {
    for (Player player : getPlayers()) {
      new Settlement(player, bge, msg).settlePlayer();
    }
  }

  private boolean dealt = false;

  private boolean active = false;

  public boolean isActive() {
    return active;
  }

  public boolean isDealt() {
    return dealt;
  }
}

File: code/shoring/blackjackEngine/src/main/kotlin/com/panopset/blackjackEngine/Cycle.kt
package com.panopset.blackjackEngine

import java.io.StringWriter

class Cycle internal constructor(
    val bge: BlackjackGameEngine,
    val strategy: Strategy
) {
    private val blackjackConfiguration: BlackjackConfiguration = bge.config
    private val countingSystems = bge.countingSystems
    val dealer = HandDealer()
    val msg: BlackjackMessages = bge.config.messages

    val players: ArrayList<Player> = ArrayList()

    fun placeBets() {
        players.clear()
        for (i in 0 until blackjackConfiguration.seats) {
            if (bge.isCountVeryNegative()) {
                bge.triggerShuffleBeforeNextHand()
            }
            val betAmount = BetAmountStrategy(bge).adjust()
            val player = Player( Wager(bge.bankroll.subtract(betAmount)))
            player.hands[0].wager.initialBet = betAmount
            players.add(player)
        }
    }

    override fun toString(): String {
        val sw = StringWriter()
        sw.append("Dealer: ")
        sw.append(String.format("%s%n", dealer.cards))
        for ((i, p) in players.withIndex()) {
            sw.append(String.format("Player %d: %s%n", i, p.toString()))
        }
        sw.append(" Bankroll:" + bge.bankroll)
        return sw.toString()
    }

    fun getActivePlayer(): Player? {
        for (p in players) {
            if (!p.isFinal) {
                return p
            }
        }
        return null
    }

    fun getRecommendedAction(): String {
        return if (dealer.isFinal || !dealer.hasCards()) {
            CMD_DEAL
        } else {
            strategy.getRecommendation(getCurrentSituation(getActivePlayerHand()))
        }
    }

    fun getStrategyLine(): StrategyLine? {
        return if (getActivePlayer() == null) {
            null
        } else {
            strategy.findStrategyLine(getCurrentSituation(getActivePlayer()!!.activeHand))
        }
    }

    val activeSituation: Situation
        get() = getCurrentSituation(getActivePlayerHand())

    fun getCurrentSituation(hand: HandPlayer?): Situation {
        return if (isDealt) {
            Situation(dealer.upCard, hand)
        } else Situation(null, hand)
    }

    fun deal() {
        placeBets()
        for (p in players) {
            p.activeHand?.dealCard(BlackjackShoe.deal(true, countingSystems))
            bge.reportNewHand()
        }
        dealer.dealCard(BlackjackShoe.deal(false, countingSystems))
        for (p in players) {
            p.activeHand?.dealCard(BlackjackShoe.deal(true, countingSystems))
        }
        dealer.dealCard(BlackjackShoe.deal(true, countingSystems))
        checkFor21()
        isDealt = true
        isActive = true
    }

    private fun checkFor21() {
        for (player in players) {
            checkFor21(player)
        }
        if (dealer.value == 21) {
            finish()
        }
    }

    private fun checkFor21(player: Player) {
        if (dealer.upCard.isAce && bge.config.isEvenMoneyOnBlackjackVace) {
            for (playerHand in player.hands) {
                if (playerHand.isNatural21()) {
                    playerHand.standWithEvenMoney()
                }
            }
        } else {
            for (playerHand in player.hands) {
                if (playerHand.value == 21) {
                    player.activeHand?.stand()
                }
            }
        }
    }

    fun getActivePlayerHand(): HandPlayer? {
        val p = getActivePlayer() ?: return null
        return p.activeHand
    }

    fun finish() {
        completeCycle()
        isActive = false
    }

    private fun completeCycle() {
        if (!dealer.isFinal) {
            prepareSettlement()
        }
    }

    private fun prepareSettlement() {
        BlackjackShoe.show(dealer.cards[0], countingSystems)
        resolvePlayerBlackjacks()
        while (!dealer.isFinal) {
            if (dealer.value < 17) {
                dealer.dealCard(bge.deal(true))
            } else if (dealer.value == 17) {
                if (blackjackConfiguration.isDealerHitSoft17 && dealer.isSoft) {
                    dealer.dealCard(bge.deal(true))
                } else {
                    dealer.stand()
                }
            } else {
                dealer.stand()
            }
        }
        bge.clearPriorHandBustedFlag()
        collectAndPayChips()
    }

    private fun resolvePlayerBlackjacks() {
        var allPlayersHaveBlackjack = true
        for (handPlayer in players) {
            if (!resolvePlayerHandsBlackjacks(handPlayer)) {
                allPlayersHaveBlackjack = false
            }
        }
        if (allPlayersHaveBlackjack) {
            dealer.stand()
        }
    }

    private fun resolvePlayerHandsBlackjacks(handPlayer: Player): Boolean {
        var allHandsAreBlackjack = true
        for (h in handPlayer.hands) {
            if (!h.isBustedOrSurrenderred()) {
                if (dealer.isNatural21() && !blackjackConfiguration.isEuropeanStyle) {
                    h.stand()
                } else {
                    if (h.isNatural21()) {
                        h.stand()
                    } else {
                        allHandsAreBlackjack = false
                    }
                }
            }
        }
        return allHandsAreBlackjack
    }

    private fun collectAndPayChips() {
        for (player in players) {
            Settlement(player, bge, msg).settlePlayer()
        }
        bge.bankroll.settle(bge.getCycle().players)
    }

    var isDealt = false
        private set
    var isActive = false
        private set

}
